File: packages/sandbox/daemon/setup/remote-golden.test.ts
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import {
  mkdtemp,
  mkdir,
  readdir,
  readFile,
  rm,
  stat,
  utimes,
  writeFile,
} from "node:fs/promises";
import { randomFillSync } from "node:crypto";
import { tmpdir } from "node:os";
import { join } from "node:path";
import type { Config } from "../types";
import {
  publishRemoteGolden,
  pruneRemoteGoldens,
  remoteGoldenPath,
  tryRestoreRemoteGolden,
} from "./remote-golden";

const CLONE_URL = "https://user:tok@github.com/acme/site.git";

function config(): Config {
  return {
    git: { repository: { cloneUrl: CLONE_URL } },
  } as unknown as Config;
}

/** A repo root with a lockfile (required — no lockfile means no cache key). */
async function makeInstallRoot(base: string, name: string): Promise<string> {
  const root = join(base, name);
  await mkdir(root, { recursive: true });
  await writeFile(join(root, "bun.lock"), '{"lockfileVersion":1}');
  return root;
}

async function seedNodeModules(root: string): Promise<void> {
  await mkdir(join(root, "node_modules", "left-pad"), { recursive: true });
  await writeFile(
    join(root, "node_modules", "left-pad", "index.js"),
    "module.exports = 1;",
  );
  // Pod-local caches — must not travel in a shared archive.
  await mkdir(join(root, "node_modules", ".vite", "deps"), { recursive: true });
  await writeFile(join(root, "node_modules", ".vite", "deps", "x"), "junk");
}

/**
 * A tree big enough that `tar -x` (many small writes) falls behind
 * `zstd -dc`, so the transfer actually has to survive backpressure.
 *
 * Incompressible content on purpose: zstd must move real bytes, not collapse
 * the whole thing into a few KB. 1500 files x 6 KB ≈ 9 MB — comfortably past
 * the ~1 MB floor below which a relayed pipe never drops anything, which is
 * exactly why the small fixtures above missed this.
 */
const BIG_FILES = 1500;
const BIG_FILE_BYTES = 6 * 1024;

async function seedBigNodeModules(root: string): Promise<void> {
  const dir = join(root, "node_modules", "bulk");
  await mkdir(dir, { recursive: true });
  await Promise.all(
    Array.from({ length: BIG_FILES }, (_, i) => {
      const buf = Buffer.alloc(BIG_FILE_BYTES);
      randomFillSync(buf);
      // Stamp the index so a truncated-but-plausible extract can't pass by
      // having the right count with the wrong contents.
      buf.write(`file-${i};`, 0);
      return writeFile(join(dir, `f${i}.bin`), buf);
    }),
  );
}

/**
 * The one archive under the store: <root>/golden/<repoHash>/<pm>-<lock>.tar.zst
 *
 * Asserts rather than destructures blindly: when publish silently produced
 * nothing, the bare `join(dir, undefined)` threw ERR_INVALID_ARG_TYPE from in
 * here, which reads as a broken helper rather than "publish failed". Combined
 * with `publishRemoteGolden` swallowing its own diagnostics when no `log` is
 * passed, a real EPIPE in the transport surfaced as a confusing TypeError.
 */
async function soleArchive(
  root: string,
): Promise<{ dir: string; path: string }> {
  const repos = await readdir(join(root, "golden"));
  expect(repos, "publish created no repo dir under the store").toHaveLength(1);
  const dir = join(root, "golden", repos[0] as string);
  const names = (await readdir(dir)).filter((n) => n.endsWith(".tar.zst"));
  expect(names, `publish left no archive in ${dir}`).toHaveLength(1);
  return { dir, path: join(dir, names[0] as string) };
}

/** Collects the transport's own failure messages so a test can report them. */
function logger(): { log: (m: string) => void; lines: string[] } {
  const lines: string[] = [];
  return { log: (m) => lines.push(m), lines };
}

const exists = (p: string) =>
  stat(p).then(
    () => true,
    () => false,
  );

let base: string;
let remoteRoot: string;
const prevRemote = process.env.GOLDEN_CACHE_REMOTE;

beforeEach(async () => {
  base = await mkdtemp(join(tmpdir(), "l2-"));
  remoteRoot = join(base, "shared");
  await mkdir(remoteRoot, { recursive: true });
  process.env.GOLDEN_CACHE_REMOTE = remoteRoot;
});

afterEach(async () => {
  if (prevRemote === undefined) delete process.env.GOLDEN_CACHE_REMOTE;
  else process.env.GOLDEN_CACHE_REMOTE = prevRemote;
  await rm(base, { recursive: true, force: true });
});

describe("remote golden (L2)", () => {
  test("publishes on one node and restores on another", async () => {
    // "Node A" installs and publishes.
    const nodeA = await makeInstallRoot(base, "node-a");
    await seedNodeModules(nodeA);
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });

    const archive = remoteGoldenPath({
      remoteRoot,
      cloneUrl: CLONE_URL,
      pm: "bun",
      lockHash: null,
    });
    // lockHash null → no key at all; the real path needs the lockfile read.
    expect(archive).toBeNull();

    // "Node B" is cold: same repo, same lockfile, empty node_modules. This is
    // the whole point of L2 — a node that never saw this repo still skips
    // the install.
    const nodeB = await makeInstallRoot(base, "node-b");
    const restored = await tryRestoreRemoteGolden({
      config: config(),
      installRoot: nodeB,
      pm: "bun",
    });

    expect(restored).toBe(true);
    expect(
      await readFile(
        join(nodeB, "node_modules", "left-pad", "index.js"),
        "utf8",
      ),
    ).toBe("module.exports = 1;");

    // Pod-local caches were excluded at publish, so they must not appear.
    expect(await exists(join(nodeB, "node_modules", ".vite"))).toBe(false);

    // No staging dir left behind.
    const leftovers = join(nodeB, `.node_modules.l2.${process.pid}`);
    expect(await exists(leftovers)).toBe(false);
  });

  test("round-trips a tree large enough to apply backpressure", async () => {
    // The regression guard for the fd-to-fd pipe. Relaying the archive through
    // this process (`p.stdout.pipe(c.stdin)`) silently dropped the tail once
    // the consumer fell behind — and BOTH children still exited 0, so the
    // truncation surfaced only later as `tar: Unexpected EOF in archive`, or
    // worse, as a permanently-published archive holding half its members.
    // Every other test here uses a ~1 KB fixture, under the size where the
    // loss ever happened; without this one a revert to the relayed pipe stays
    // green.
    const nodeA = await makeInstallRoot(base, "big-a");
    await seedBigNodeModules(nodeA);
    const pub = logger();
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
      log: pub.log,
    });
    // Surface the transport's own message. Without this a broken pipe or a
    // failed read-back shows up only as "no archive", one layer too late.
    expect(pub.lines.filter((l) => !l.includes("published"))).toEqual([]);

    // Publish must not declare success on a truncated archive.
    const { path } = await soleArchive(remoteRoot);
    expect(await exists(path)).toBe(true);

    const nodeB = await makeInstallRoot(base, "big-b");
    expect(
      await tryRestoreRemoteGolden({
        config: config(),
        installRoot: nodeB,
        pm: "bun",
      }),
    ).toBe(true);

    // Exact count, not "some files" — the observed failure delivered 532 of
    // 1069 members and looked fine at a glance.
    const restored = await readdir(join(nodeB, "node_modules", "bulk"));
    expect(restored.length).toBe(BIG_FILES);

    // Spot-check both ends of the stream: a dropped tail is what this catches,
    // so the LAST member matters most.
    for (const i of [0, BIG_FILES - 1]) {
      const buf = await readFile(
        join(nodeB, "node_modules", "bulk", `f${i}.bin`),
      );
      expect(buf.length).toBe(BIG_FILE_BYTES);
      expect(buf.subarray(0, `file-${i};`.length).toString()).toBe(
        `file-${i};`,
      );
    }
  }, 60_000);

  test("does nothing when the shared store is not configured", async () => {
    // The discriminating case: identical setup, feature off. Without this the
    // test above would pass even if the code ignored the env entirely.
    const nodeA = await makeInstallRoot(base, "off-a");
    await seedNodeModules(nodeA);
    delete process.env.GOLDEN_CACHE_REMOTE;

    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });
    const nodeB = await makeInstallRoot(base, "off-b");
    expect(
      await tryRestoreRemoteGolden({
        config: config(),
        installRoot: nodeB,
        pm: "bun",
      }),
    ).toBe(false);
    expect(await exists(join(nodeB, "node_modules"))).toBe(false);
  });

  test("a truncated archive fails closed, leaving no partial tree", async () => {
    // The failure that would be worst: a half-extracted node_modules that
    // later code reads as complete, so the boot skips install and breaks
    // somewhere unrelated. Restore must leave nothing behind.
    const nodeA = await makeInstallRoot(base, "trunc-a");
    await seedNodeModules(nodeA);
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });

    const nodeB = await makeInstallRoot(base, "trunc-b");
    const { path } = await soleArchive(remoteRoot);
    await writeFile(path, "not a zstd stream");

    expect(
      await tryRestoreRemoteGolden({
        config: config(),
        installRoot: nodeB,
        pm: "bun",
      }),
    ).toBe(false);
    expect(await exists(join(nodeB, "node_modules"))).toBe(false);
    expect(await exists(join(nodeB, `.node_modules.l2.${process.pid}`))).toBe(
      false,
    );
  });

  test("publish is idempotent and does not rewrite an existing archive", async () => {
    const nodeA = await makeInstallRoot(base, "idem-a");
    await seedNodeModules(nodeA);
    const opts = { config: config(), installRoot: nodeA, pm: "bun" };
    await publishRemoteGolden(opts);

    const { dir, path } = await soleArchive(remoteRoot);
    const first = await stat(path);

    // A second healthy boot on another node must not rewrite the archive —
    // concurrent publishers across the fleet would otherwise churn the
    // shared store on every boot.
    await publishRemoteGolden(opts);
    expect((await stat(path)).mtimeMs).toBe(first.mtimeMs);

    // And no temp file survived.
    expect((await readdir(dir)).filter((e) => e.includes(".tmp."))).toEqual([]);
  });

  test("prune bounds the store by TTL and per-repo cap", async () => {
    const nodeA = await makeInstallRoot(base, "gc-a");
    await seedNodeModules(nodeA);
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });
    const { dir, path: fresh } = await soleArchive(remoteRoot);

    // Three older archives (one already past the TTL) plus an in-flight temp
    // from a concurrent publisher on another node.
    const day = 24 * 60 * 60 * 1000;
    const aged: string[] = [];
    for (const [i, ageDays] of [1, 2, 9].entries()) {
      const p = join(dir, `bun-old${i}.tar.zst`);
      await writeFile(p, "x");
      const t = new Date(Date.now() - ageDays * day);
      await utimes(p, t, t);
      aged.push(p);
    }
    const inflight = join(dir, "bun-new.tar.zst.tmp.999");
    await writeFile(inflight, "half an archive");

    await pruneRemoteGoldens(remoteRoot, { maxPerRepo: 2 });

    // Newest 2 by mtime survive the cap; the 9-day-old one is past the TTL and
    // would be dropped even under a generous cap.
    expect(await exists(fresh)).toBe(true);
    expect(await exists(aged[0] as string)).toBe(true);
    expect(await exists(aged[1] as string)).toBe(false);
    expect(await exists(aged[2] as string)).toBe(false);
    // A temp file is another node's in-flight publish, not a prunable entry —
    // reaping it would corrupt a publish already in progress.
    expect(await exists(inflight)).toBe(true);
  });

  test("a successful publish prunes the store", async () => {
    // The wiring, not the rule: without this the store still grows forever
    // however correct pruneRemoteGoldens is on its own.
    const nodeA = await makeInstallRoot(base, "gcw-a");
    await seedNodeModules(nodeA);
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });
    const { dir } = await soleArchive(remoteRoot);

    const stale = join(dir, "bun-stale.tar.zst");
    await writeFile(stale, "x");
    const old = new Date(Date.now() - 9 * 24 * 60 * 60 * 1000);
    await utimes(stale, old, old);

    // A new lockfile is a new key, so this publish is not the idempotent no-op.
    await writeFile(join(nodeA, "bun.lock"), '{"lockfileVersion":2}');
    await publishRemoteGolden({
      config: config(),
      installRoot: nodeA,
      pm: "bun",
    });

    expect(await exists(stale)).toBe(false);
    // Both real keys are recent, so the cap keeps them.
    expect((await readdir(dir)).length).toBe(2);
  });
});

File: packages/sandbox/daemon-go/internal/setup/remotegolden_test.go
package setup

// Real files, real tar/zstd, no mocks — the whole point of this tier is that a
// tree survives a compress/decompress round trip through a store the daemon does
// not control, and a mocked pipe proves nothing about that.

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"
)

// requireTools skips rather than fails when the host lacks zstd: the sandbox
// image ships it, a contributor's machine may not, and a red suite there would
// train people to ignore it.
func requireTools(t *testing.T) {
	t.Helper()
	for _, bin := range []string{"tar", "zstd"} {
		if _, err := exec.LookPath(bin); err != nil {
			t.Skipf("%s not on PATH", bin)
		}
	}
}

// installRootWithTree builds an install root holding a node_modules with enough
// shape to prove the round trip: nested dirs, a file per dir, and one of the
// pod-local caches that must NOT travel.
func installRootWithTree(t *testing.T, pkgs int) string {
	t.Helper()
	root := t.TempDir()
	if err := os.WriteFile(filepath.Join(root, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	for i := 0; i < pkgs; i++ {
		dir := filepath.Join(root, "node_modules", fmt.Sprintf("pkg-%d", i), "dist")
		if err := os.MkdirAll(dir, 0o755); err != nil {
			t.Fatal(err)
		}
		body := fmt.Sprintf("module.exports = %d;", i)
		if err := os.WriteFile(filepath.Join(dir, "index.js"), []byte(body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	cache := filepath.Join(root, "node_modules", ".vite")
	if err := os.MkdirAll(cache, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(cache, "junk"), []byte("pod-local"), 0o644); err != nil {
		t.Fatal(err)
	}
	return root
}

func params(remoteRoot, installRoot string) RemoteGoldenParams {
	return RemoteGoldenParams{
		RemoteRoot:  remoteRoot,
		CloneUrl:    "https://github.com/acme/site.git",
		InstallRoot: installRoot,
		Pm:          "bun",
	}
}

func TestRemoteGoldenArchivePath(t *testing.T) {
	t.Run("keyed by repo, pm and lockfile", func(t *testing.T) {
		got := RemoteGoldenArchivePath("/store", "https://github.com/acme/site.git", "bun", "abc")
		want := filepath.Join("/store", "golden", repoCacheKey("https://github.com/acme/site.git"), "bun-abc.tar.zst")
		if got != want {
			t.Fatalf("got %q want %q", got, want)
		}
	})
	t.Run("credentials do not change the key", func(t *testing.T) {
		bare := RemoteGoldenArchivePath("/store", "https://github.com/acme/site.git", "bun", "abc")
		creds := RemoteGoldenArchivePath("/store", "https://user:tok@github.com/acme/site.git", "bun", "abc")
		if bare != creds {
			t.Fatalf("credential-stripping broken:\n bare  %q\n creds %q", bare, creds)
		}
	})
	t.Run("empty when a component is missing", func(t *testing.T) {
		for _, c := range [][4]string{
			{"", "url", "bun", "hash"},
			{"/store", "", "bun", "hash"},
			{"/store", "url", "bun", ""},
		} {
			if got := RemoteGoldenArchivePath(c[0], c[1], c[2], c[3]); got != "" {
				t.Fatalf("expected empty for %v, got %q", c, got)
			}
		}
	})
	t.Run("two repos never share an archive", func(t *testing.T) {
		a := RemoteGoldenArchivePath("/store", "https://github.com/acme/a.git", "bun", "same")
		b := RemoteGoldenArchivePath("/store", "https://github.com/acme/b.git", "bun", "same")
		if a == b {
			t.Fatal("distinct repos collided on one archive — bun does not re-verify cache content")
		}
	})
}

func TestRemoteEnabled(t *testing.T) {
	t.Setenv(remoteEnabledEnvVar, "")
	if RemoteEnabled() {
		t.Fatal("must be off when unset")
	}
	t.Setenv(remoteEnabledEnvVar, "/golden-cache")
	if !RemoteEnabled() {
		t.Fatal("must be on when set to a path")
	}
}

// The tier's whole purpose: a tree published where one pod ran is restored where
// another one runs, with no node-local state in common.
func TestRemoteGoldenPublishOnOneNodeRestoreOnAnother(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeA := installRootWithTree(t, 12)
	PublishRemoteGolden(params(store, nodeA))

	// A cold node: same lockfile, no node_modules at all.
	nodeB := t.TempDir()
	if err := os.WriteFile(filepath.Join(nodeB, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	if !TryRestoreRemoteGolden(params(store, nodeB)) {
		t.Fatal("restore reported no hit on a cold node")
	}

	got, err := os.ReadFile(filepath.Join(nodeB, "node_modules", "pkg-7", "dist", "index.js"))
	if err != nil {
		t.Fatalf("restored tree missing a package: %v", err)
	}
	if string(got) != "module.exports = 7;" {
		t.Fatalf("restored contents wrong: %q", got)
	}
	// Pod-local caches must not travel in a shared archive.
	if _, err := os.Stat(filepath.Join(nodeB, "node_modules", ".vite")); err == nil {
		t.Fatal(".vite travelled in the archive — every consumer pays for pod-local churn")
	}
	// No staging left behind.
	entries, _ := os.ReadDir(nodeB)
	for _, e := range entries {
		if len(e.Name()) > 15 && e.Name()[:15] == ".node_modules.l" {
			t.Fatalf("staging dir survived: %s", e.Name())
		}
	}
}

// A tree big enough that the transfer cannot complete inside one pipe buffer —
// this is the case that caught a truncating implementation in the TypeScript
// version, where a 1 KB fixture passed and a multi-MB one did not.
func TestRemoteGoldenRoundTripsUnderBackpressure(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeA := installRootWithTree(t, 40)
	// ~8 MB of incompressible payload, well past any pipe buffer.
	big := make([]byte, 8<<20)
	for i := range big {
		big[i] = byte(i * 7)
	}
	if err := os.WriteFile(filepath.Join(nodeA, "node_modules", "pkg-0", "dist", "big.bin"), big, 0o644); err != nil {
		t.Fatal(err)
	}
	PublishRemoteGolden(params(store, nodeA))

	nodeB := t.TempDir()
	if err := os.WriteFile(filepath.Join(nodeB, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	if !TryRestoreRemoteGolden(params(store, nodeB)) {
		t.Fatal("restore reported no hit")
	}
	back, err := os.ReadFile(filepath.Join(nodeB, "node_modules", "pkg-0", "dist", "big.bin"))
	if err != nil {
		t.Fatal(err)
	}
	if len(back) != len(big) {
		t.Fatalf("payload truncated: got %d bytes want %d", len(back), len(big))
	}
	for i := range back {
		if back[i] != big[i] {
			t.Fatalf("payload corrupted at byte %d", i)
		}
	}
}

// Discriminating counterpart: without this, the tests above would pass even if
// the code ignored its configuration and restored from somewhere else.
func TestRemoteGoldenDormantWhenUnconfigured(t *testing.T) {
	requireTools(t)
	store := t.TempDir()

	t.Setenv(remoteEnabledEnvVar, "")
	nodeA := installRootWithTree(t, 4)
	PublishRemoteGolden(params("", nodeA))
	if entries, _ := os.ReadDir(store); len(entries) != 0 {
		t.Fatal("published with the tier disabled")
	}

	nodeB := t.TempDir()
	if err := os.WriteFile(filepath.Join(nodeB, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	if TryRestoreRemoteGolden(params("", nodeB)) {
		t.Fatal("restored with the tier disabled")
	}
	if _, err := os.Stat(filepath.Join(nodeB, "node_modules")); err == nil {
		t.Fatal("created node_modules with the tier disabled")
	}
}

// A truncated archive must fail closed. A partial node_modules that later code
// reads as complete is worse than a miss: the boot skips install and fails
// somewhere unrelated.
func TestRemoteGoldenTruncatedArchiveFailsClosed(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeA := installRootWithTree(t, 20)
	PublishRemoteGolden(params(store, nodeA))

	archive := RemoteGoldenArchivePath(store, "https://github.com/acme/site.git", "bun",
		LockfileHash(nodeA, "bun"))
	full, err := os.ReadFile(archive)
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(archive, full[:len(full)/2], 0o644); err != nil {
		t.Fatal(err)
	}

	nodeB := t.TempDir()
	if err := os.WriteFile(filepath.Join(nodeB, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	if TryRestoreRemoteGolden(params(store, nodeB)) {
		t.Fatal("reported success on a truncated archive")
	}
	if _, err := os.Stat(filepath.Join(nodeB, "node_modules")); err == nil {
		t.Fatal("left a partial node_modules behind")
	}
	entries, _ := os.ReadDir(nodeB)
	for _, e := range entries {
		if len(e.Name()) > 15 && e.Name()[:15] == ".node_modules.l" {
			t.Fatalf("staging dir survived a failed restore: %s", e.Name())
		}
	}
}

// Publish no-ops once the key exists. It matters because the store has no
// rename: a second publisher rewriting a live key would expose a partial object
// to concurrent readers.
func TestRemoteGoldenPublishIsIdempotent(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeA := installRootWithTree(t, 6)
	PublishRemoteGolden(params(store, nodeA))
	archive := RemoteGoldenArchivePath(store, "https://github.com/acme/site.git", "bun",
		LockfileHash(nodeA, "bun"))
	first, err := os.Stat(archive)
	if err != nil {
		t.Fatal(err)
	}

	// Change the tree so a rewrite would be observable, then publish again.
	if err := os.WriteFile(filepath.Join(nodeA, "node_modules", "pkg-0", "dist", "index.js"),
		[]byte("module.exports = 'rewritten';"), 0o644); err != nil {
		t.Fatal(err)
	}
	PublishRemoteGolden(params(store, nodeA))

	second, err := os.Stat(archive)
	if err != nil {
		t.Fatal(err)
	}
	if !first.ModTime().Equal(second.ModTime()) || first.Size() != second.Size() {
		t.Fatal("republished over an existing key")
	}
}

// Garbage at the key — not a truncated archive but never-valid bytes — is also a
// miss, not a crash and not a partial tree.
//
// NOTE: this exercises the RESTORE side only. Publish's read-back guard (drop the
// object when `zstd -dc | tar -tf -` rejects what was just written) is not
// covered: publish no-ops on an existing key, so a corrupt object cannot be
// planted ahead of it, and making zstd emit a bad archive on demand would mean
// faking the pipe, which defeats the point of testing against real tools.
func TestRemoteGoldenRestoreRejectsCorruptArchive(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeB := t.TempDir()
	if err := os.WriteFile(filepath.Join(nodeB, "bun.lock"), []byte("lockfile-v1"), 0o644); err != nil {
		t.Fatal(err)
	}
	archive := RemoteGoldenArchivePath(store, "https://github.com/acme/site.git", "bun",
		LockfileHash(nodeB, "bun"))
	if err := os.MkdirAll(filepath.Dir(archive), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(archive, []byte("not a zstd stream"), 0o644); err != nil {
		t.Fatal(err)
	}

	if TryRestoreRemoteGolden(params(store, nodeB)) {
		t.Fatal("restored from a corrupt archive")
	}
	if _, err := os.Stat(filepath.Join(nodeB, "node_modules")); err == nil {
		t.Fatal("left a node_modules behind after a corrupt archive")
	}
}

func TestPruneRemoteGoldens(t *testing.T) {
	now := time.Date(2026, 8, 6, 12, 0, 0, 0, time.UTC)
	write := func(t *testing.T, dir, name string, age time.Duration) string {
		t.Helper()
		if err := os.MkdirAll(dir, 0o755); err != nil {
			t.Fatal(err)
		}
		p := filepath.Join(dir, name)
		if err := os.WriteFile(p, []byte("x"), 0o644); err != nil {
			t.Fatal(err)
		}
		stamp := now.Add(-age)
		if err := os.Chtimes(p, stamp, stamp); err != nil {
			t.Fatal(err)
		}
		return p
	}

	t.Run("absent store is a no-op", func(t *testing.T) {
		pruneRemoteGoldens("", GoldenTTL, GoldenMaxPerRepo, now, nil)
		pruneRemoteGoldens(filepath.Join(t.TempDir(), "absent"), GoldenTTL, GoldenMaxPerRepo, now, nil)
	})

	t.Run("drops archives past the TTL", func(t *testing.T) {
		store := t.TempDir()
		repo := filepath.Join(store, "golden", "repo-a")
		fresh := write(t, repo, "bun-fresh.tar.zst", time.Hour)
		stale := write(t, repo, "bun-stale.tar.zst", 8*24*time.Hour)
		pruneRemoteGoldens(store, GoldenTTL, GoldenMaxPerRepo, now, nil)
		if _, err := os.Stat(fresh); err != nil {
			t.Fatal("pruned an archive inside the TTL")
		}
		if _, err := os.Stat(stale); err == nil {
			t.Fatal("kept an archive past the TTL")
		}
	})

	t.Run("keeps only the newest per repo", func(t *testing.T) {
		store := t.TempDir()
		repo := filepath.Join(store, "golden", "repo-a")
		var paths []string
		for i := 0; i < GoldenMaxPerRepo+3; i++ {
			paths = append(paths, write(t, repo, fmt.Sprintf("bun-%d.tar.zst", i), time.Duration(i)*time.Minute))
		}
		pruneRemoteGoldens(store, GoldenTTL, GoldenMaxPerRepo, now, nil)
		for i, p := range paths {
			_, err := os.Stat(p)
			if i < GoldenMaxPerRepo && err != nil {
				t.Fatalf("pruned archive %d, inside the cap", i)
			}
			if i >= GoldenMaxPerRepo && err == nil {
				t.Fatalf("kept archive %d, past the cap", i)
			}
		}
	})

	t.Run("leaves non-archive entries alone", func(t *testing.T) {
		store := t.TempDir()
		repo := filepath.Join(store, "golden", "repo-a")
		other := write(t, repo, "README", 8*24*time.Hour)
		pruneRemoteGoldens(store, GoldenTTL, GoldenMaxPerRepo, now, nil)
		if _, err := os.Stat(other); err != nil {
			t.Fatal("pruned something that is not an archive")
		}
	})

	t.Run("one repo's cap does not affect another", func(t *testing.T) {
		store := t.TempDir()
		a := filepath.Join(store, "golden", "repo-a")
		b := filepath.Join(store, "golden", "repo-b")
		for i := 0; i < GoldenMaxPerRepo+2; i++ {
			write(t, a, fmt.Sprintf("bun-%d.tar.zst", i), time.Duration(i)*time.Minute)
		}
		only := write(t, b, "bun-only.tar.zst", time.Minute)
		pruneRemoteGoldens(store, GoldenTTL, GoldenMaxPerRepo, now, nil)
		if _, err := os.Stat(only); err != nil {
			t.Fatal("pruned another repo's only archive")
		}
	})
}

// A successful publish prunes, so the store stays bounded wherever it is
// writable without anything having to walk it on a schedule.
func TestRemoteGoldenPublishPrunes(t *testing.T) {
	requireTools(t)
	store := t.TempDir()
	t.Setenv(remoteEnabledEnvVar, store)

	nodeA := installRootWithTree(t, 4)
	repoDir := filepath.Dir(RemoteGoldenArchivePath(store, "https://github.com/acme/site.git", "bun",
		LockfileHash(nodeA, "bun")))
	if err := os.MkdirAll(repoDir, 0o755); err != nil {
		t.Fatal(err)
	}
	stale := filepath.Join(repoDir, "bun-ancient.tar.zst")
	if err := os.WriteFile(stale, []byte("x"), 0o644); err != nil {
		t.Fatal(err)
	}
	old := time.Now().Add(-30 * 24 * time.Hour)
	if err := os.Chtimes(stale, old, old); err != nil {
		t.Fatal(err)
	}

	PublishRemoteGolden(params(store, nodeA))

	if _, err := os.Stat(stale); err == nil {
		t.Fatal("publish did not prune a stale archive")
	}
}
